sys/include/net/gcoap.h

```diff
@@ -1045,6 +1046,320 @@ ssize_t gcoap_encode_link(const coap_resource_t *resource, char *buf,
 sock_dtls_t *gcoap_get_sock_dtls(void);
 #endif
 
+/**
+ * @defgroup net_gcoap_nanocoap gCoAP packet abstraction layer
+ * @ingroup net_gcoap
+ *
+ * gCoAP utilizes @ref net_nanocoap to parse and abstract CoAP packets. To separate those to modules
+ * cleanly, this module provides wrappers around nanocoap's functionality.
+ * @{
+ */
+/**
+ * @name    Buffer manipulation
+ * @{
+ */
+static inline uint8_t *gcoap_pkt_get_buf(coap_pkt_t *pdu)
+{
+    return (uint8_t *)pdu->hdr;
+}
+
+static inline void gcoap_pkt_set_buf(coap_pkt_t *pdu, uint8_t *buf)
+{
+    pdu->hdr = (coap_hdr_t *)buf;
+}
+/** @} */
+
+/**
+ * @name    Header manipulation
+ * @{
+ */
+static inline int gcoap_pkt_parse(coap_pkt_t *pdu, uint8_t *buf, size_t len)
+{
+    return coap_parse(pdu, buf, len);
+}
+
+static inline ssize_t gcoap_pkt_build_hdr(uint8_t *buf, unsigned type, uint8_t *token,
+                                          size_t token_len, unsigned code, uint16_t id)
+{
+    return coap_build_hdr((coap_hdr_t *)buf, type, token, token_len, code, id);
+}
+
+static inline void gcoap_pkt_init(coap_pkt_t *pdu, uint8_t *buf, size_t buf_len, size_t hdr_len)
+{
+    coap_pkt_init(pdu, buf, buf_len, hdr_len);
+}
+
+static inline unsigned gcoap_pkt_get_type(coap_pkt_t *pdu)
+{
+    return coap_get_type(pdu);
+}
+
+static inline void gcoap_pkt_set_type(coap_pkt_t *pdu, unsigned type)
+{
+    coap_hdr_set_type(pdu->hdr, type);
+}
+
+static inline void gcoap_pkt_set_token_len(coap_pkt_t *pdu, uint8_t token_len)
+{
+    coap_set_token_len(pdu, token_len);
+}
+
+static inline unsigned gcoap_pkt_get_code_raw(coap_pkt_t *pdu)
+{
+    return coap_get_code_raw(pdu);
+}
+
+static inline unsigned gcoap_pkt_get_code_class(coap_pkt_t *pdu)
+{
+    return coap_get_code_class(pdu);
+}
+
+static inline unsigned gcoap_pkt_get_code_detail(coap_pkt_t *pdu)
+{
+    return coap_get_code_detail(pdu);
+}
+
+static inline void gcoap_pkt_set_code_raw(coap_pkt_t *pdu, uint8_t code)
+{
+    coap_hdr_set_code(pdu->hdr, code);
+}
+
+static inline coap_method_flags_t gcoap_pkt_get_method_flag(coap_pkt_t *pdu)
+{
+    assert(coap_get_code_class(pdu) == COAP_CLASS_REQ);
+    return coap_method2flag(coap_get_code_detail(pdu));
+}
+
+static inline uint16_t gcoap_pkt_get_id(coap_pkt_t *pdu)
+{
+    return coap_get_id(pdu);
+}
+
+static inline bool gcoap_pkt_has_id(coap_pkt_t *pdu, uint16_t id)
+{
+    return coap_get_id(pdu) == id;
+}
+
+static inline ssize_t gcoap_pkt_get_token(coap_pkt_t *pdu, uint8_t *token)
+{
+    ssize_t res = coap_get_token_len(pdu);
+    if (res > 0) {
+        memcpy(token, pdu->token, res);
+    }
+    return res;
+}
+
+static inline bool gcoap_pkt_has_token(coap_pkt_t *pdu, const uint8_t *token, size_t token_len)
+{
+    return (coap_get_token_len(pdu) == token_len) && (memcmp(pdu, token, token_len) == 0);
```

Comment:
How does that `memcmp` work on `coap_pkt_t`?

Reply:
Oops, I think I forgot to add the member here ;-).